sklearn/utils/_array_api.py

```diff
@@ -357,6 +386,48 @@ def _expit(X):
     return 1.0 / (1.0 + xp.exp(-X))
 
 
+def _add_to_diagonal(array, value, xp):
+    # Workaround for the lack of support for xp.reshape(a, shape, copy=False) in
+    # numpy.array_api: https://github.com/numpy/numpy/issues/23410
+    value = xp.asarray(value, dtype=array.dtype)
+    if _is_numpy_namespace(xp):
+        array_np = numpy.asarray(array)
+        array_np.flat[:: array.shape[0] + 1] += value
+        return xp.asarray(array_np)
+    elif value.ndim == 1:
+        for i in range(array.shape[0]):
+            array[i, i] += value[i]
+    else:
+        # scalar value
+        for i in range(array.shape[0]):
+            array[i, i] += value
+
+
+def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
+    if xp is None:
+        xp, _ = get_namespace(sample_score)
+    if normalize and _is_numpy_namespace(xp):
+        sample_score_np = numpy.asarray(sample_score)
+        if sample_weight is not None:
+            sample_weight_np = numpy.asarray(sample_weight)
+        else:
+            sample_weight_np = None
+        return float(numpy.average(sample_score_np, weights=sample_weight_np))
+
+    if normalize:
+        if sample_weight is not None:
+            scale = xp.sum(sample_weight)
+        else:
+            scale = sample_score.shape[0]
+        if scale != 0:
+            sample_score = sample_score / scale
+
+    if sample_weight is not None:
+        return float(sample_score @ sample_weight)
+    else:
+        return float(xp.sum(sample_score))
```

Comment:
Note: we need to call `float(...)` to convert the output to Python scalar instead of 0-dim Array objects.

The `accuracy_score` docstring states that the return value is of type `float` but is as actually a numpy 0-dim array which might be confusing.

Converting to float is probably saner for user code. Otherwise, some consumer code would need to explicitly move back the scalar code to host memory explicitly to manipulate 0-dim Array instance back from GPU memory if when the input data comes from GPU device memory.

Reply:
I don't know how this fits into the bigger picture of what scikit-learn is doing, but it's worth noting that calling `float()` would prevent things from working with libraries that use lazy evaluation. There aren't currently any such libraries conforming to the array API, but there likely will be in the future. See the discussion at https://github.com/data-apis/array-api/issues/642.

Reply:
That's a good point. I will add a note for now. I think we need a dedicated issue in scikit-learn to explore how to use lazy Array API implementations with scikit-learn estimators.

Reply:
I opened the following issue to track lazy evaluation:

- #26724